source/versioned-api/versioned-api.rst

```diff
@@ -0,0 +1,273 @@
+
+Drivers SHOULD group the ``serverApi`` option with other similar client options
+like ``autoEncryptionOpts``.
+
+
+ServerApiVersion enumeration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This specification and subsequent specifications will define the known API
+versions. Drivers SHOULD define an enumeration containing the known API
+versions, using the version identifiers given. Drivers MAY deviate from the
+version identifiers used in this and subsequent specifications if doing so is
+necessary given the driver's programming language's constraints.
+
+Drivers for languages that don't have enums (e.g. PHP) MUST expose the version
+as a string, but SHOULD offer constants to allow for IDE features such as code
+completion. In these cases, the driver MUST validate (e.g. when the application
+provides a version string to the ``ServerApi`` class) that the version string is
+valid and trigger a client-side error if an unknown API version was used.
+
+
+ServerApi class
+~~~~~~~~~~~~~~~
+
+The ``ServerApi`` class stores an API version, along with flags that decide
+whether or not unknown or deprecated commands in the specified API version
+trigger a server-side error. A ``version`` MUST be specified when declaring an
+API version, while the ``strict`` and ``deprecationErrors`` options are both
+optional and default to ``false``. The ``ServerApi`` class is considered
+immutable; changes to the declared API version MUST be prohibited.
+
+
+Declared Version Inheritance
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Drivers MUST ensure that users cannot override the API version declared in the
+``MongoClient`` instance. This includes the ``MongoDatabase`` and
+``MongoCollection`` classes, as well as any operations in these classes. See the
+rationale for more details.
+
+
+Sending Declared API Version to the Server
+------------------------------------------
+
+The declared API version MUST be sent to the server with every command, with the
+exception of the cases listed below. Drivers MUST NOT use a server's reported
+``maxWireVersion`` to decide whether it supports the versioned API. The server
+will reply with an error if the declared API version is not supported. If the
+user does not declare an API version, the driver MUST NOT send any API
+versioning options to the server.
+
+
+Command Syntax
+~~~~~~~~~~~~~~
+
+The options from the declared API version are mapped to the following command
+options:
+
+===================== ========================
+**ServerApi field**   **Command option**
+``version``           ``apiVersion``
+``strict``            ``apiStrict``
+``deprecationErrors`` ``apiDeprecationErrors``
+===================== ========================
+
+If an API version was declared, drivers MUST add the ``apiVersion`` option to
+every command that is sent to a server. Drivers MUST add the ``apiStrict`` and
+``apiDeprecationErrors`` options if they were specified by the user, even when
+the specified value is equal to the server default. Drivers SHOULD NOT add these
```

Comment:
can we make this "SHOULD NOT" a "MUST NOT"? Or is there a circumstance where it might be necessary to send a value the user didn't specify?

Reply:
I used "SHOULD NOT" to allow drivers to use a `bool` for `strict` and `deprecationErrors` without tracking whether the user supplied it. If I change this to "MUST NOT", I'd also remove the default values for the above options from the `ServerApi` class to ensure that they are empty by default, which indicates that the driver must not add them.

Reply:
I think unless we wish to deviate from the server's default values for these, we should just not send them if the user didn't specify them. This would be consistent with how we handle other options that we forward along to the server.

Reply:
On line 160 what does "these" refer to? strict & deprecation errors or all 3 options?

Reply:
I am not following the reason for should vs must but regarding this:

>  I'd also remove the default values for the above options from the ServerApi class to ensure that they are empty by default, which indicates that the driver must not add them.

We already require drivers not sending default values for `multi` and `upsert` in crud. Therefore I expect it should be possible for drivers to not send default values here.

Reply:
Changed "SHOULD NOT" to "MUST NOT" and reworded the last sentence for clarity:
> Drivers MUST NOT add any API versioning options if the user did not specify them.